packages/client/src/pages/Game/towers/Tower.tsx

```diff
@@ -68,124 +109,284 @@ class Tower {
+  };
+
+  public drawTowerRange(
+    context: CanvasRenderingContext2D = this.engine.context!.build!,
+  ) {
+    context.beginPath();
+    context.lineWidth = 1;
+    // context.setLineDash([10, 15]);
+    context.fillStyle = this.towerParams.strokeStyle;
     // draw tower range
-    this.engine.context?.arc(
-      this.currentPosition.x - this.towerParams.width / 2,
-      this.currentPosition.y - this.towerParams.height / 2,
+    context.arc(
+      this.currentPosition.x - this.towerParams.baseWidth / 2,
+      this.currentPosition.y - this.towerParams.baseHeight / 2,
       this.towerParams.attackRange,
       0,
       360,
     );
-    this.engine.context?.stroke();
-    this.engine.context?.closePath();
-  }
-
-  public drawTower() {
-    // draw tower 2d representation
-    this.draw();
-    // draw tower range
-    if (this.engine.isCanBuild) {
-      this.drawTowerRange();
-    }
+    context.fill();
+    context.closePath();
   }
 
   public isEnemyInRange(enemy: Enemy) {
     const xDistance =
-      enemy.currentPosition.x - this.currentPosition.x + this.towerParams.width;
+      this.currentPosition.x -
+      this.towerParams.baseWidth / 2 -
+      (enemy.currentPosition.x + enemy.enemyParams.width! / 2);
     const yDistance =
-      enemy.currentPosition.y -
-      this.currentPosition.y +
-      this.towerParams.height;
+      this.currentPosition.y -
+      this.towerParams.baseHeight / 2 -
+      (enemy.currentPosition.y + enemy.enemyParams.height! / 2);
     if (Math.hypot(xDistance, yDistance) < this.towerParams.attackRange) {
       this.target = enemy;
+      if (!this.attackIntervalTimer) {
+        this.setAttackInterval();
+      }
       return true;
     }
     return false;
   }
 
   public findTarget() {
-    if (this.engine.enemies!.length) {
-      if (!this.target) {
-        this.engine.enemies?.forEach((enemy) => {
+    if (!this.engine.enemies!.length) return;
+    if (!this.target) {
+      this.engine.enemies?.forEach((enemy) => {
+        // check range only for the closest enemies
+        if (
+          Math.abs(
+            this.currentPosition.x -
+              this.towerParams.baseWidth / 2 -
+              enemy.currentPosition.x +
+              enemy.enemyParams.width! / 2,
+          ) < this.towerParams.attackRange &&
+          Math.abs(
+            this.currentPosition.y -
+              this.towerParams.baseHeight / 2 -
+              enemy.currentPosition.y +
+              enemy.enemyParams.height! / 2,
+          ) < this.towerParams.attackRange
+        ) {
           return this.isEnemyInRange(enemy);
-        });
-      } else {
-        if (!this.isEnemyInRange(this.target)) {
-          this.target = null;
         }
+        return false;
+      });
+    } else {
+      if (!this.isEnemyInRange(this.target)) {
+        this.target = null;
       }
     }
   }
 
-  public findTargetVector() {
+  public findTargetAngle() {
     // there is no spoon, Neo
```

Comment:
I choose the red one, Mr.Smith

Reply:
It's actually not a vector, it's an angle. I'm getting target vector into Projectile class, but later it was a vector.